### PR TITLE
Check Manual downloads

### PR DESCRIPTION
**Pull request by harsh020**
## Part 2 of 2 | Check if file was manually downloaded

PR for issue #2422

Tasks -

* [x] Check if file was downloaded manually
* [ ] Write `.INFO` file

### CODE REVIEW

File: tensorflow_datasets/core/download/download_manager.py [hunk 1]
```diff
@@ -444,6 +459,14 @@ def _download(self, resource: Union[str, resource_lib.Resource]):
       resource = resource_lib.Resource(url=resource)
     url = resource.url
 
+    manually_downloaded_path = self._check_manually_downloaded(url)
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/download/download_manager.py:462]: Could you add a test in `download_manager_test.py` ?

File: tensorflow_datasets/core/download/download_manager.py [hunk 2]
```diff
@@ -421,6 +421,21 @@ def _find_existing_path(self, url: str, url_path: str) -> Optional[str]:
...
+    url_info = self._url_infos.get(url)
+    # If url info was recorded with filename
+    manually_downloaded_path = None
+    if url_info and url_info.filename:
+        manually_downloaded_path = os.path.join(
+          self._manual_dir,
+          url_info.filename
+        )
+        # File not manually downloaded
+        if not tf.io.gfile.exists(manually_downloaded_path):
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/download/download_manager.py:435]: Could you also check that the checksums match `url_info.checksums` if `_force_checksums_validation` is True ?

**Review by Conchylicultor — CHANGES_REQUESTED** [review]
Thank you @harsh020 , this looks great!

File: tensorflow_datasets/core/download/download_manager.py (hunk 2 above)
Comment (harsh020) [reply tensorflow_datasets/core/download/download_manager.py]: Hi, I have implemented the checksum check for the case when the user hasn't extracted the dataset. I'm working on the case when the user has extracted the dataset.

Thanks.

**Comment by harsh020** [comment]
@Conchylicultor, I was wondering don't we need to also check if the file is extracted (`tf.io.gfile.exist(filename without extension)`) as the `url_info.filename` will just have the `basename` with extension.

**Review by Conchylicultor — APPROVED** [review]
Thank you! Merging this soon :)

**Comment by Conchylicultor** [comment]
> @Conchylicultor, I was wondering don't we need to also check if the file is extracted (i.e, `tf.io.gfile.exist(filename without extension)`, maybe the user downloaded and extracted the dataset in the `manual_dir`) as the `url_info.filename` will just have the `basename` with extension.
> 
> Thanks.

I don't think we want the user to manually extract the data because:

1. This add additional work for the user
2. This creates additional complexity as the script might want to explicitly control the extraction (e.g. if using `dl_manager.iter_archive`, or with nested archives)

**Comment by harsh020** [comment]
@Conchylicultor The `INFO` files are still left to write. Should I do it in the next commit?

Thanks.

**Comment by Conchylicultor** [comment]
> @Conchylicultor The `INFO` files are still left to commit (the commit is all set to push). Should I do it in the next commit?
> 
> Thanks.

For manual downloads, I don't think we need `INFO` files. Those are mainly to track automated downloads.

**Pull request merged**